### PR TITLE
125 cmd hook

### PR DESCRIPTION
**Pull request by jamesroutley**

### CODE REVIEW

File: docs/stack_config.rst [hunk 1]
```diff
@@ -348,6 +348,26 @@ Example:
...
+********
+
+Executes string as a shell or PowerShell command.
+
+Syntax:
+
+.. code-block:: yaml
+
+    <hook_point>:
+      - !cmd <bash_command>
```
Comment (theseanything) [inline docs/stack_config.rst]: I assume was renamed to show that it support Windows commands....

in which case we should remove `bash` from `<bash_command>`

File: docs/stack_config.rst [hunk 2]
```diff
@@ -348,6 +348,26 @@ Example:
...
+
+Example:
```
Comment (theseanything) [inline docs/stack_config.rst]: Maybe add windows example?

File: sceptre/hooks/cmd.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import subprocess
+from sceptre.hooks import Hook
+from sceptre.exceptions import InvalidHookArgumentTypeError
+
+
+class Cmd(Hook):
+    """
+    Cmd implements a Sceptre hook which can run arbitrary shell commands.
```
Comment (theseanything) [inline sceptre/hooks/cmd.py]: again is shell a linux only term?

File: sceptre/hooks/cmd.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(Cmd, self).__init__(*args, **kwargs)
+
+    def run(self):
+        """
+        Runs the argument string in child process.
```
Comment (theseanything) [inline sceptre/hooks/cmd.py]: Runs the argument string in *sub* process.

File: tests/test_hooks/test_cmd.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    def test_run_with_non_str_argument(self):
+        self.cmd.argument = None
+        with pytest.raises(InvalidHookArgumentTypeError):
+            self.cmd.run()
+
+    @patch('sceptre.hooks.bash.subprocess.check_call')
+    def test_run_with_str_argument(self, mock_call):
+        self.cmd.argument = u"echo hello"
+        self.cmd.run()
+        mock_call.assert_called_once_with(['/bin/bash', '-c', "echo hello"])
```
Comment (theseanything) [inline tests/test_hooks/test_cmd.py]: Maybe add another test with a command give a non-zero exit?

**Comment by jamesroutley** [comment]
Agree with changes, I'll add them tomorrow.

File: tests/test_hooks/test_cmd.py (hunk 5 above)
Comment (jamesroutley) [inline tests/test_hooks/test_cmd.py]: added below

**Comment by jamesroutley** [comment]
@seanrankine added those changes

File: sceptre/hooks/cmd.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+        :raises: sceptre.exceptions.InvalidTaskArgumentTypeException
+        :raises: subprocess.CalledProcessError
+        """
+        if not isinstance(self.argument, basestring):
+            raise InvalidHookArgumentTypeError(
+                'The argument "{0}" is the wrong type - cmd hooks require '
+                'arguments of type string.'.format(self.argument)
+            )
+        subprocess.check_call(["/bin/bash", "-c", self.argument])
```
Comment (theseanything) [inline sceptre/hooks/cmd.py]: Remove the /bin/bash -c

Comment (theseanything) [inline sceptre/hooks/cmd.py]: Remove the /bin/bash -c

File: sceptre/hooks/cmd.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+    def run(self):
+        """
+        Runs the argument string in a subprocess.
+
+        :raises: sceptre.exceptions.InvalidTaskArgumentTypeException
+        :raises: subprocess.CalledProcessError
+        """
+        try:
+            subprocess.check_call(self.argument, shell=True)
```
Comment (theseanything) [inline sceptre/hooks/cmd.py]: Sorry, again we did we opt for `shell=True`?

Comment (jamesroutley) [inline sceptre/hooks/cmd.py]: Is that 'did we' or 'why did we'?

- did we: yes
- why did we: it's gives us access to constructs defined at the shell level, such as piping. It's considered a security risk if you pass unfiltered external input to `subprocess.check_call` but we decided it was alright here because it'd only run in an environment where the user already had access.

Comment (theseanything) [inline sceptre/hooks/cmd.py]: Ahah sorry - meant "why did we". So what about build servers? Someone could push and execute potentially malicious code. If it were `shell=False`, do we still have the ability to invoke a shell via `/bin/bash -c`?

Comment (theseanything) [inline sceptre/hooks/cmd.py]: Decided parsing hook arguments would be difficult and possible security risks with sceptre aren't solved by shell=False.
